sources/osgShadow/ShadowFrustumIntersection.js

```diff
@@ -0,0 +1,204 @@
+                    return; // culled
+            }
+
+            // light cull
+            if ( this._lightFrustum.getCurrentMask() !== 0 ) {
+                // father bounding sphere is not totally inside
+                // now test this one
+                if ( !this._lightFrustum.containsBoundingSphere( bs ) )
+                    return; // culled
+            }
+
+            this._cameraFrustum.pushCurrentMask();
+            this._lightFrustum.pushCurrentMask();
+
+            this.pushMatrix( matrix );
+
+            this.traverse( transform );
+
+            this._cameraFrustum.popCurrentMask();
+            this._lightFrustum.popCurrentMask();
+
+            this.popMatrix();
+        },
+        applyBoundingBox: ( function () {
+            var tmpVec = Vec3.create();
+            var tmpCorner = Vec3.create();
+            return function ( bbox ) {
+                var stackLength = this._matrixStack.length;
+                var matrix = this._matrixStack[ stackLength - 1 ];
+                if ( Matrix.isIdentity( matrix ) ) {
+                    this._bb.expandByBoundingBox( bbox );
+                } else if ( bbox.valid() ) {
+                    this._bb.expandByVec3( Matrix.transformVec3( matrix, bbox.corner( 0, tmpCorner ), tmpVec ) );
+                    this._bb.expandByVec3( Matrix.transformVec3( matrix, bbox.corner( 1, tmpCorner ), tmpVec ) );
+                    this._bb.expandByVec3( Matrix.transformVec3( matrix, bbox.corner( 2, tmpCorner ), tmpVec ) );
+                    this._bb.expandByVec3( Matrix.transformVec3( matrix, bbox.corner( 3, tmpCorner ), tmpVec ) );
+                    this._bb.expandByVec3( Matrix.transformVec3( matrix, bbox.corner( 4, tmpCorner ), tmpVec ) );
+                    this._bb.expandByVec3( Matrix.transformVec3( matrix, bbox.corner( 5, tmpCorner ), tmpVec ) );
+                    this._bb.expandByVec3( Matrix.transformVec3( matrix, bbox.corner( 6, tmpCorner ), tmpVec ) );
+                    this._bb.expandByVec3( Matrix.transformVec3( matrix, bbox.corner( 7, tmpCorner ), tmpVec ) );
+                }
+            };
+        } )(),
+
+        apply: function ( node ) {
+
+            var typeID = node.getTypeID();
+
+            if ( typeID === MatrixTransform.getTypeID() ) {
+                this.applyTransform( node );
+                return;
+
+            } else if ( typeID === Geometry.getTypeID() ) {
+                var bs = this._bs;
+                var matrix = this._matrixStack[ this._matrixStack.length - 1 ];
+                Matrix.transformBoundingSphere( matrix, node.getBound(), bs );
+
+                // camera cull
+                if ( this._cameraFrustum.getCurrentMask() !== 0 ) {
+                    // father bounding sphere is not totally inside
+                    // now test this one
+                    if ( !this._cameraFrustum.containsBoundingSphere( bs ) )
+                        return; // culled
```

Comment:
could we talk about tihs logic. It looks like if light or camera is culled we stop. I guess it's good but that would be great to add extra comments to explain the overall logic about it